### PR TITLE
Feature/par 44 sim translate simulation script

### PR DESCRIPTION
**Pull request by Mqqrc**
Run conversation with flight booking mock api. Works with gpt-4. Json fixer is being finished in a new branch so gpt 3.5 is not completely working yet.

### CODE REVIEW

File: src/simulation/db/models/simulation.model.ts [hunk 1]
```diff
@@ -20,7 +21,8 @@ const SimulationSchema: Schema = new Schema(
     scenario: { type: String, enum: Object.values(SimulationScenario), required: true },
     type: { type: String, enum: Object.values(ConversationType), required: true },
     domain: { type: String, enum: Object.values(ConversationDomain), required: true },
-    agents: [{ type: Schema.Types.ObjectId, ref: 'Agent' }],
+    userAgent: { type: Schema.Types.ObjectId, ref: 'Agent' },
+    serviceAgent: { type: Schema.Types.ObjectId, ref: 'Agent' },
```
Comment (weirdkey) [inline src/simulation/db/models/simulation.model.ts:25]: Something that we need to consider is that when running A/B testing, then we would have two service agents. We could either handle it as an array or handle it as two objects. @ardaakcabuyuk should give his opinion on how to handle this.

**Comment by weirdkey** [comment]
Fantastic job. Well done 👏 . One comment from my side except the one above. Can you remove the debug console logs which are comments?

**Review by ardaakcabuyuk — CHANGES_REQUESTED** [review]
Very good job guys, thank you a lot for your effort! I'm waiting for your comments 🚀

**Review by ardaakcabuyuk — COMMENTED** [review]
I'll resolve the conflicts, you don't need to worry about them.

File: src/simulation/db/models/simulation.model.ts (hunk 1 above)
Comment (Mqqrc) [reply src/simulation/db/models/simulation.model.ts:25]: Yeah exactly, we agree with Arda. But let's implement that later when it's needed.

File: src/simulation/agents/custom.agent.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import moment from 'moment';
 import { appendFileSync } from 'fs';
 import { MsgTypes } from '../db/enum/enums';
 
-class MsgHistoryItem {
+export class MsgHistoryItem {
```
[the remark this answers is not in the document]
Comment (R0N222) [reply src/simulation/agents/custom.agent.ts:9]: Honestly not sure if it to much sense, as most of the things stored in the MsgHistoryItem are only used in very special cases. We can store the data far more efficient in the database, also it uses langchain components like the BaseMessage.
image

Marc will take care of storing the message data efficiently in his flushing to database task.
Note MsgHistoryItem is also a class and not an interface like MessageDocument.

File: src/simulation/agents/custom.agent.ts [hunk 3]
```diff
@@ -67,8 +67,8 @@ type ToolDescription = {
   output?: string;
 };
 
-class CustomAgent {
-  chat_model: BaseChatModel;
+export class CustomAgent {
```
[the remark this answers is not in the document]
Comment (Mqqrc) [reply src/simulation/agents/custom.agent.ts:70]: We should talk about that in a quick meeting, as this is also part of my current task. Have you got time tomorrow? Just suggest a time on slack.

File: src/simulation/agents/custom.agent.ts [hunk 4]
```diff
@@ -107,28 +107,37 @@ class CustomAgent {
...
+      const responseMessage = await this.chatModel.call(messages);
 
-      if (action === MsgTypes.MSGTOUSER) {
+      //console.log('MESSAGE: \n' + responseMessage.content.toString() + '\n\n');
+      response = this.getFixedJson(responseMessage.content.toString());
+      action = response.action;
+      actionInput = response.action_input ?? {};
+
+      //console.log("Message: check 1")
+      if (action === MsgTypes.MSGTOUSER || action == 'message_to_user' || action == 'None') {
```
[the remark this answers is not in the document]
Comment (R0N222) [reply src/simulation/agents/custom.agent.ts:140]: The AI always messes up to correctly print MSGTOUSER, and often prints 'message_to_user', so it is an easy way to fix a large chunk of formatting issues.

[the remark this answers is not in the document]
Comment (weirdkey) [reply src/simulation/agents/custom.agent.ts:135]: I think it is part of the LLM basic styling conventions to handle it as in Python and it is OK to leave it that way

**Review by ardaakcabuyuk — APPROVED** [review]

**Pull request merged**